torch/lib/THD/base/init_methods/InitMethod.cpp

```diff
@@ -1,43 +1,59 @@
 #include "InitMethod.hpp"
 
+#ifdef THD_INIT_EXTENSION_H
+#define INCF(F) INCF_(F)
+#define INCF_(F) #F
+#include INCF(THD_INIT_EXTENSION_H)
+#endif
+
 namespace thd {
 namespace init {
 
-InitMethod::Config initTCP(std::string argument, rank_type world_size,
-                           std::string group_name, int rank);
-InitMethod::Config initFile(std::string argument, rank_type world_size,
-                            std::string group_name, int rank);
-InitMethod::Config initEnv(int world_size, std::string group_name, int rank);
-
-}
+InitMethod::Config initTCP(std::string argument,
+                           int world_size_r,
+                           std::string group_name,
+                           int rank);
+InitMethod::Config initFile(std::string argument,
+                            int world_size_r,
+                            std::string group_name,
+                            int rank);
+InitMethod::Config initEnv(std::string argument,
+                           int world_size_r,
+                           std::string group_name,
+                           int rank);
+
+InitMethodFuncMap initMethods({
+    {"env://", ::thd::init::initEnv},
+    {"file://", ::thd::init::initFile},
+    {"tcp://", ::thd::init::initTCP}
+
+#ifdef THD_INIT_EXTENSION_H
+    ,
+    /**
+     * Additional method pairs can be defined in THD_INIT_EXTENSION_H header
+     * to extend the init methods
+     */
+    THD_INIT_EXTENSION_METHODS
```

Comment:
Ok, so I had a slightly different mechanism in mind. I think it would be much cleaner to just have a function like this:
```
void registerInitMethod(std::string name, std::function<::thd::InitMethod::Config(std::string, int, std::string, int)> method)
```
that would register the code at run time. The users will have to insert an extra call to actually register the method before they use it, but I think that's much easier to understand than an optionally defined macro. Wdyt?

Reply:
For now, I actually prefer not to let the user register anything at runtime, instead, the extension header/init_method pair would be pretty sufficient for the extension and it's very simple to use (see the other diff I sent, define a pair would be enough). Plus, we will most likely only have this one extension case (very unlikely to have another) since all current init_methods are pretty sufficient for the public users IMO.